### PR TITLE
properly handle `IndexCreate` markers in rocksdb recovery

### PR DESCRIPTION
**Pull request by gnusi**
Please note that for legal reasons we require you to sign the 
[Contributor Agreement](https://www.arangodb.com/documents/cla.pdf)
before we can accept your pull requests.

### CODE REVIEW

File: arangod/IResearch/IResearchView.cpp [hunk 1]
```diff
@@ -316,6 +316,117 @@ inline void insertDocument(
...
+    LOG_TOPIC(DEBUG, arangodb::iresearch::IResearchFeature::IRESEARCH)
+      << "registering consolidation policy '" << size_t(entry.type()) << "' with IResearch view '" << viewName << "' run id '" << size_t(&runId) << " segment threshold '" << entry.segmentThreshold() << "' segment count '" << segmentCount.load() << "'";
+
+    try {
+      writer.consolidate(entry.policy(), false);
+      forceCommit = true; // a consolidation policy was found requiring commit
+    } catch (std::exception const& e) {
+      LOG_TOPIC(WARN, arangodb::iresearch::IResearchFeature::IRESEARCH)
+        << "caught exception during registeration of consolidation policy '" << size_t(entry.type()) << "' with IResearch view '" << viewName << "': " << e.what();
+      IR_EXCEPTION();
```
Comment (danielhlarkin) [inline arangod/IResearch/IResearchView.cpp]: `IR_EXCEPTION` is a deceptive name. I assumed this threw another exception, but it seems to just log something. Maybe consider renaming it in the future? `IR_LOG_EXCEPTION()`?

File: arangod/IResearch/IResearchRocksDBRecoveryHelper.cpp [hunk 2]
```diff
@@ -42,12 +43,74 @@
...
+  }
+
+  return std::make_pair(vocbase, vocbase->lookupCollection(pair.second));
+}
+
+std::vector<std::shared_ptr<arangodb::Index>> lookupLinks(
+    arangodb::LogicalCollection& coll
+) {
+  // FIXME better to have
+  // LogicalCollection::getIndexes(std::vector<>&)
```
[the remark this answers is not in the document]
Comment (gnusi) [reply arangod/IResearch/IResearchRocksDBRecoveryHelper.cpp]: wrong signature, I meant `visitor`

File: arangod/IResearch/IResearchRocksDBRecoveryHelper.cpp [hunk 3]
```diff
@@ -42,12 +43,74 @@
...
+  auto indexes = col->getIndexes();
+
+  auto it = std::find_if(
+    indexes.begin(), indexes.end(),
+    [iid](std::shared_ptr<arangodb::Index> const& idx) {
+      return idx->id() == iid && idx->type() == arangodb::Index::IndexType::TRI_IDX_TYPE_IRESEARCH_LINK;
+  });
+
+  // TODO FIXME find a better way to retrieve an iResearch Link
+  // cannot use static_cast/reinterpret_cast since Index is not related to IResearchLink
```
[the remark this answers is not in the document]
Comment (gnusi) [reply arangod/IResearch/IResearchRocksDBRecoveryHelper.cpp]: don't know better way as for now

File: arangod/IResearch/IResearchRocksDBRecoveryHelper.cpp [hunk 4]
```diff
@@ -42,12 +43,74 @@
 #include "VocBase/LogicalView.h"
 #include "VocBase/vocbase.h"
 
-using namespace arangodb;
-using namespace arangodb::iresearch;
+NS_LOCAL
```
[the remark this answers is not in the document]
Comment (gnusi) [reply arangod/IResearch/IResearchRocksDBRecoveryHelper.cpp]: done

File: arangod/IResearch/IResearchRocksDBRecoveryHelper.cpp [hunk 5]
```diff
@@ -42,12 +43,74 @@
...
+
+  // TODO FIXME find a better way to retrieve an iResearch Link
+  // cannot use static_cast/reinterpret_cast since Index is not related to IResearchLink
+
+  return it == indexes.end()
+    ? nullptr
+    : dynamic_cast<arangodb::iresearch::IResearchLink*>(it->get());
+}
+
+NS_END
```
[the remark this answers is not in the document]
Comment (gnusi) [reply arangod/IResearch/IResearchRocksDBRecoveryHelper.cpp]: done

File: arangod/IResearch/IResearchRocksDBRecoveryHelper.cpp [hunk 6]
```diff
@@ -42,12 +43,74 @@
...
+
+using namespace arangodb;
+using namespace arangodb::iresearch;
```
[the remark this answers is not in the document]
Comment (gnusi) [reply arangod/IResearch/IResearchRocksDBRecoveryHelper.cpp]: done

File: arangod/IResearch/IResearchView.cpp (hunk 1 above)
Comment (gnusi) [reply arangod/IResearch/IResearchView.cpp]: will fix it in iresearch

File: arangod/IResearch/IResearchRocksDBRecoveryHelper.h [hunk 7]
```diff
@@ -62,15 +62,39 @@ class IResearchRocksDBRecoveryHelper : public RocksDBRecoveryHelper {
...
+    TRI_idx_iid_t indexId,
+    TRI_voc_cid_t viewId
+  );
+
+  void ensureLink(
+    TRI_voc_tick_t dbId,
+    TRI_voc_cid_t cid,
+    arangodb::velocypack::Slice indexSlice
+  );
+
```
[the remark this answers is not in the document]
Comment (gnusi) [reply arangod/IResearch/IResearchRocksDBRecoveryHelper.h]: the only downside I see is the fact that I have to make `IndexId` public which not a big problem I suppose

**Pull request merged**